src/rax.c

```diff
@@ -456,21 +483,38 @@ raxNode *raxCompressNode(raxNode *n, unsigned char *s, size_t len, raxNode **chi
  * means that the current node represents the key (that is, none of the
  * compressed node characters are needed to represent the key, just all
  * its parents nodes). */
-static inline size_t raxLowWalk(rax *rax, unsigned char *s, size_t len, raxNode **stopnode, raxNode ***plink, int *splitpos, raxStack *ts) {
+static inline size_t raxLowWalk(
+    rax *rax, unsigned char *s, size_t len, raxNode **stopnode, raxNode ***plink, int *splitpos, raxStack *ts,
+    uint8_t *stopoffset
+) {
     raxNode *h = rax->head;
     raxNode **parentlink = &rax->head;
 
     size_t i = 0; /* Position in the string. */
     size_t j = 0; /* Position in the node children (or bytes if compressed).*/
+    size_t k = 0; /* Current child offset or mismatch offset */
     while(h->size && i < len) {
         debugnode("Lookup current node",h);
         unsigned char *v = h->data;
-
         if (h->iscompr) {
             for (j = 0; j < h->size && i < len; j++, i++) {
                 if (v[j] != s[i]) break;
             }
             if (j != h->size) break;
+        } else if (stopoffset) { /* Scan preserving stopoffset */
+            for (j = 0; j < h->size; j++) {
+                if (v[j] >= s[i]) break;
+            }
+            if (v[j] > s[i]) {
+                k = j;
+                j = h->size;
+                break;
+            }
+            if (j == h->size) {
+                k = j;
+                break;
+            }
+            i++;
```

Comment:
It looks like we can just modify the else branch.
Btw, I'm not sure why the original implementation iterates over all child nodes, when the node is not compressed, it must be ordered, so when `v[j] >= s[i]`, it can exit early.
```c
else {
    /* Even when h->size is large, linear scan provides good
     * performances compared to other approaches that are in theory
     * more sounding, like performing a binary search. */
    int found = 0;
    for (j = 0; j < h->size; j++) {
        if (v[j] == s[i]) {
            found = 1;
            break;
        } else if (v[j] > s[i]) {
            break;
        }
    }
    if (!found) break;
    i++;
}
```

Reply:
Do you mean modify the else branch in the current PR incorporating all commits? I can try that and test it.

Reply:
I mean we can merge the ` else if (stopoffset)` and `else` branches.

Reply:
Ah - but your proposal will not set k, the current node's child_offset, which is needed by raxSeek.

Reply:
In fact `j` is the original `k`

Reply:
I still do not see how it will work - perhaps you should propose a complete raxLowWalk and I will incorporate and test it.

Reply:
I see now. I have just put in a 1st iteration which passes all tests, now I will test your proposal (iteration 2) and then check the performance of each. I seem to recall that antirez chose his simple scan because it was faster than any alternative - I tested this briefly quite a while ago and found it seemed true - but we'll see how things go this time!

Reply:
I tested the 2nd iteration, and it passed all tests, and is equally as fast as Iteration 1 on numeric keys, including when iterating. However it is consistently slower than the original code and Iteration 1 when benchmarking alphanumeric keys. Trees built with alphanumeric keys have shorter spans, which probably accounts for the difference.